### ISSUE TITLE
Nextcloud OCR

### ISSUE THREAD

**Issue by wakielo** [alexbelgium/hassio-addons#71]
21:36:43] INFO: Installing OCR
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
WARNING: The repository tag for world dependency 'tesseract-ocr@community' does not exist
ERROR: Not committing changes due to missing repository tags. Use --force-broken-world to override.
[cont-init.d] 51-ocr: exited 99.

Hi Alex, I install the addon and cant get the OCR working... Do you have any idea of what I am doing wrong?

**Comment by alexbelgium** [comment]
Hi, thanks for reporting. I'll check

**Comment by alexbelgium** [comment]
Hi, I've seen the issue. A new version is pushed that will solve the issue.

Have fun with the addon!

**Comment by wakielo** [comment]
I think it's fixed now! But now I'm struggling how nextcloud index my
files, may I enable  something?

**Comment by alexbelgium** [comment]
Not sure to understand... If you mean for ocr it is a bit more complex as you need to install an nextcloud app and everything. Here is the description : https://markus-blog.de/index.php/2018/06/21/how-to-install-fulltextsearch-in-nextcloud-13-with-elasticsearch-and-tesseract-ocr/

My addon only performs the step 2, as I think other steps can be done from within the app ui...

**Comment by wakielo** [comment]
I keep trying understand the problem with elasticsearch... 

Your addon index all of your content trought elasticsearch automatically??

**Comment by alexbelgium** [comment]
ah... well not automatically, it still needs to be installed also... I'll try to find a way to install it automatically at addon install

**Comment by wakielo** [comment]
I remove and installed again the apps and doesn't index my files... Neither
it doesn't let me start the OCR process

**Comment by alexbelgium** [comment]
Actually indeed an instance of elasticsearch needs to be connected with the elasticsearch for Full Text Search app... And it is quite complex to deploy on rpi... I've created a test addon for that but I crashes on launch due to lack of memory. An alternative is to use elasticsearch cloud but there is only a 30d free trial period

**Comment by wakielo** [comment]
I tried starting a node but it was imposible. I tried to follow according this guide https://help.nextcloud.com/t/how-to-get-elasticsearch-working/80926/10 and adapt the script to our environment but toooo many things to hardcode...

For what purpose your nextcloud uses the OCR?

**Comment by alexbelgium** [comment]
I tried to do the same without success so far... 

On my side l created an addon based on an elasticsearch docker image. It nearly works except for a single issue : it needs to have an increased max ram which is not possible as the base system is read only in HA. You are welcome to install my "elasticsearch DO NOT Use" addon if you have any idea on how to make it work! It will start but then stop on initialization due to this setting. Once the addon is running it would be "just" a matter of referring the addon ip and port in the elasticsearch full text search app of nextcloud to use it. 

Initially I had put tesseract-ocr to enable this, but now I realize the only way to achieve real ocr in the app is either to use elasticsearch cloud (paid service) or have a preexisting elasticsearch installation running which is quite complex. So I've just renamed the addon from "nextcloud ocr" to just "nextcloud" ;)

**Comment by alexbelgium** [comment]
Actually the script you found is great. I extracted from it an env variable that could help my elasticsearch addon run. If it works, it would simply be a matter of creating new options in the nextcloud addon to refer its url and port and it would install and configure it (thanks to your script) . Once I'll have access to an amd64 machine I'll try again.

**Comment by alexbelgium** [comment]
Thanks to your script, I believe it works! Please install and run my elasticsearch container. When you go in nextcloud and in the full text search tab of the parameters, you'll see that the server is correctly recognized and auto filled! We still need to do some tests but it's looking on the right path!

**Comment by wakielo** [comment]
could you make it work? 

I have tried the elastic plugin and it partially works but it ask me to install ingest-attachment, go to install it the container you upload and crash when finish the download... :(

**Comment by alexbelgium** [comment]
Hi, I'll test on my system. Where did it request ingest-attachment? I hadn't seen it in the logs? It's going to work, just a matter of time :)

**Comment by alexbelgium** [comment]
Let's do it sequentially. I'll update a bit this post as I advance.

1. Setting elasticsearch
- I've installed the addon "elasticsearch do not use" and run it without any settings. There are some errors in the logs, but the addon still start
- I went in terminal and typed `curl 192.168.178.69:9200` . This provided me with a json response showing that elastic search is installed and works properly. Step 1 done :-)
- !!! The addon is not fully developed and I believe that it will not save its database upon addon restart. But let's work on that once it will already work ;-)

2. Setting nextcloud
- I've installed the apps "Full Text search", Full "Text Search elasticsearch" and "Full Text search files"
- When going in the settings, I click on Full Text Search, select "elasticsearch" as the platform and the elasticsearch engine prefills.

**Comment by wakielo** [comment]
Oh Alex you advanced too much!

I'm trying handle with it from fresh install and I get this to access to nextcloud...

`
Composer autoloader not found, unable to continue. Check the folder "3rdparty". Running "git submodule update --init" will initialize the git submodule that handles the subfolder "3rdparty".`

**Comment by alexbelgium** [comment]
Mmh that doesn't look good... You get that when running nextcloud? Could you post the full log please?

**Comment by wakielo** [comment]
To support LSIO projects visit:
https://www.linuxserver.io/donate/
-------------------------------------
GID/UID
-------------------------------------
User uid:    1000
User gid:    1000
-------------------------------------
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-config: executing... 
[cont-init.d] 20-config: exited 0.
[cont-init.d] 30-define_user: executing... 
[09:41:01] INFO: Checking permissions...
[09:41:01] INFO: ... done
[cont-init.d] 30-define_user: exited 0.
[cont-init.d] 30-keygen: executing... 
[09:41:02] INFO: No ssl certificates set. Auto generating ones...
Generating a RSA private key
...........................+++++
...............................+++++
writing new private key to '/ssl/nextcloud/keys/cert.key'
-----
... adding ssl certs in files
[09:41:02] INFO: ... done
[cont-init.d] 30-keygen: exited 0.
[cont-init.d] 40-config: executing... 
[cont-init.d] 40-config: exited 0.
[cont-init.d] 50-install: executing... 
[cont-init.d] 50-install: exited 0.
[cont-init.d] 51-elasticsearch: executing... 
touch: cannot touch '/var/scripts/fetch_lib.sh': No such file or directory
[cont-init.d] 51-elasticsearch: exited 1.
[cont-init.d] 51-ocr: executing... 
Installing OCR
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/3) Installing giflib (5.2.1-r0)
(2/3) Installing leptonica (1.81.1-r0)
(3/3) Installing tesseract-ocr (4.1.1-r5)
Executing busybox-1.33.1-r3.trigger
OK: 321 MiB in 220 packages
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/main/x86_64/APKINDEX.tar.gz
fetch http://dl-cdn.alpinelinux.org/alpine/v3.14/community/x86_64/APKINDEX.tar.gz
(1/1) Installing tesseract-ocr-data-spa (4.1.1-r5)
OK: 338 MiB in 221 packages
[09:41:07] INFO: OCR Language installed : spa
[cont-init.d] 51-ocr: exited 0.
[cont-init.d] 60-memcache: executing... 
[cont-init.d] 60-memcache: exited 0.
[cont-init.d] 61-trusted_domains: executing... 
[09:41:08] INFO: Currently set trusted domains :
/var/run/s6/etc/cont-init.d/61-trusted_domains: line 14: /data/config/www/nextcloud/occ: Is a directory
[09:41:08] INFO: No trusted domain set yet. The first one will be set when doing initial configuration
[09:41:08] INFO: Trusted domains set in the configuration. Refreshing domains.
[09:41:08] INFO: ... removing previously added trusted domain (except for first one created)
/var/run/s6/etc/cont-init.d/61-trusted_domains: line 24: /data/config/www/nextcloud/occ: Is a directory
[cont-init.d] 61-trusted_domains: exited 126.
[cont-init.d] 70-aliases: executing... 
[cont-init.d] 70-aliases: exited 0.
[cont-init.d] 90-custom-folders: executing... 
[cont-init.d] 90-custom-folders: exited 0.
[cont-init.d] 99-custom-files: executing... 
[custom-init] no custom files found exiting...
[cont-init.d] 99-custom-files: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Could not open input file: /data/config/www/nextcloud/cron.php
Could not open input file: /data/config/www/nextcloud/cron.php
Could not open input file: /data/config/www/nextcloud/cron.php
Could not open input file: /data/config/www/nextcloud/cron.php
Could not open input file: /data/config/www/nextcloud/cron.php
Could not open input file: /data/config/www/nextcloud/cron.php

/var/run/s6/etc/cont-init.d/61-trusted_domains: line 14: /data/config/www/nextcloud/occ: Is a directory


[09:41:08] INFO: No trusted domain set yet. The first one will be set when doing initial configuration


[09:41:08] INFO: Trusted domains set in the configuration. Refreshing domains.


[09:41:08] INFO: ... removing previously added trusted domain (except for first one created)


/var/run/s6/etc/cont-init.d/61-trusted_domains: line 24: /data/config/www/nextcloud/occ: Is a directory


[cont-init.d] 61-trusted_domains: exited 126.


[cont-init.d] 70-aliases: executing... 


[cont-init.d] 70-aliases: exited 0.


[cont-init.d] 90-custom-folders: executing... 


[cont-init.d] 90-custom-folders: exited 0.


[cont-init.d] 99-custom-files: executing... 


[custom-init] no custom files found exiting...


[cont-init.d] 99-custom-files: exited 0.


[cont-init.d] done.


[services.d] starting services


[services.d] done.
```

I get that when fresh install the new container of nextcloud

**Comment by alexbelgium** [comment]
That's weird, I haven't changed anything to the nextcloud main addon! I'll check on my system

**Comment by alexbelgium** [comment]
For me a clean install of nexcloud runs fine...

Here is my log
`[cont-init.d] 00-banner.sh: exited 0.
[cont-init.d] 01-envfile: executing... 
[cont-init.d] 01-envfile: exited 0.
[cont-init.d] 10-adduser: executing... 
-------------------------------------
          _         ()
         | |  ___   _    __
         | | / __| | |  /  \
         | | \__ \ | | | () |
         |_| |___/ |_|  \__/
Brought to you by linuxserver.io
-------------------------------------
To support LSIO projects visit:
https://www.linuxserver.io/donate/
-------------------------------------
GID/UID
-------------------------------------
User uid:    1000
User gid:    1000
-------------------------------------
[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 20-config: executing... 
[cont-init.d] 20-config: exited 0.
[cont-init.d] 30-define_user: executing... 
[10:31:10] INFO: Checking permissions...
[10:31:10] INFO: ... the user has changed from : to 1000:1000 ...
Creating possible missing Directories
chmod Files and Directories.  This could take some time, please wait...
chown Directories. This could take some time, please wait...
chmod/chown .htaccess
[10:31:10] INFO: ... done
[cont-init.d] 30-define_user: exited 0.
[cont-init.d] 30-keygen: executing... 
[10:31:11] INFO: No ssl certificates set. Auto generating ones...
Generating a RSA private key
...+++++
.............................................+++++
writing new private key to '/ssl/nextcloud/keys/cert.key'
-----
... adding ssl certs in files
[10:31:11] INFO: ... done
[cont-init.d] 30-keygen: exited 0.
[cont-init.d] 40-config: executing... 
[cont-init.d] 40-config: exited 0.
[cont-init.d] 50-install: executing... 
[cont-init.d] 50-install: exited 0.
[cont-init.d] 51-elasticsearch: executing... 
touch: cannot touch '/var/scripts/fetch_lib.sh': No such file or directory
[cont-init.d] 51-elasticsearch: exited 1.
[cont-init.d] 51-ocr: executing... 
[cont-init.d] 51-ocr: exited 0.
[cont-init.d] 60-memcache: executing... 
[cont-init.d] 60-memcache: exited 0.
[cont-init.d] 61-trusted_domains: executing... 
[10:31:42] INFO: Currently set trusted domains :
Nextcloud is not installed - only a limited number of commands are available
                                                                   
  There are no commands defined in the "config:system" namespace.  
                                                                   
[10:31:43] INFO: No trusted domain set yet. The first one will be set when doing initial configuration
[10:31:43] INFO: Trusted domains set in the configuration. Refreshing domains.
[10:31:43] INFO: ... removing previously added trusted domain (except for first one created)
Nextcloud is not installed - only a limited number of commands are available
                                                                   
  There are no commands defined in the "config:system" namespace.  
                                                                   
[cont-init.d] 61-trusted_domains: exited 1.
[cont-init.d] 70-aliases: executing... 
[cont-init.d] 70-aliases: exited 0.
[cont-init.d] 90-custom-folders: executing... 
[cont-init.d] 90-custom-folders: exited 0.
[cont-init.d] 99-custom-files: executing... 
[custom-init] no custom files found exiting...
[cont-init.d] 99-custom-files: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
Exception: Not installed in /data/config/www/nextcloud/lib/base.php:277
Stack trace:
#0 /data/config/www/nextcloud/lib/base.php(649): OC::checkInstalled()
#1 /data/config/www/nextcloud/lib/base.php(1083): OC::init()
#2 /data/config/www/nextcloud/cron.php(42): require_once('/data/config/ww...')
#3 {main}`

**Comment by wakielo** [comment]
Let me try again... I did something wrong

**Comment by alexbelgium** [comment]
The key think is that after having set the initial database, you need to restart the addon to set the trusted domains and other scripts... I haven't found a way to automatize this yet

**Comment by wakielo** [comment]
I think i got it but I cant make any search yet... I'll keep you in touch

**Comment by wakielo** [comment]
I got it... but it doesnt search into my pdfs....

**Comment by alexbelgium** [comment]
mmh... your pdf files are searchable? for me it worked directly. Or do they need OCR ?

**Comment by wakielo** [comment]
No, just testing with pdf from  nextcloud's default install...

**Comment by alexbelgium** [comment]
Weird for me it had worked...

**Comment by wakielo** [comment]
It might be a matter of resources on the machine where I run the HA...

**Comment by wakielo** [comment]
Forget it...! I disable this plugin "Files - Tesseract OCR" and it worked like a charm!

**Comment by alexbelgium** [comment]
Sorry, I'll keep it open to track the advance here https://github.com/alexbelgium/hassio-addons/issues/71#issuecomment-961464467 if ok for you :-)

**Comment by wakielo** [comment]
Ah ok don't worry, I thought you would prefer keep it close

**Comment by alexbelgium** [comment]
Thanks it should be all good now. I've implemented the automatizatio mn scripts with two new options : enable full text search (which installs all relevant nextcloud apps), and the possibility to encode the elasticsearch server in the addon options.

Have a nice day!
